File: ios/App/Sources/ChapbookAppModel/CatalogSession.swift
import Chapbook
import Foundation

/// One open catalog, on the one thread it is allowed to be on.
///
/// Every `Catalog` call blocks and the handle is one thread's at a time,
/// so a session owns a serial queue and hops every call onto it. The
/// catalog is opened on that queue on first use, and the origin's
/// credential is set on it before every fetch — read fresh each time, so
/// a sign-in stored while a feed was open reaches the next request.
public actor CatalogSession {
    private let transport: HTTPTransport
    private let credentials: Credentials
    private let queue = DispatchSerialQueue(label: "chapbook-catalog")
    private var catalog: Catalog?

    public nonisolated var unownedExecutor: UnownedSerialExecutor { queue.asUnownedSerialExecutor() }

    public init(transport: HTTPTransport, credentials: Credentials) {
        self.transport = transport
        self.credentials = credentials
    }

    /// Run a block over the catalog, credentialed for `url`'s origin.
    public func use<T: Sendable>(for url: URL, _ body: @Sendable (Catalog) throws -> T) throws -> T {
        let catalog = try open()
        try catalog.setAuthorization(credentials.authorization(for: url))
        return try body(catalog)
    }

    private func open() throws -> Catalog {
        if let catalog { return catalog }
        let opened = try Catalog(transport: transport)
        catalog = opened
        return opened
    }
}

File: android/app/src/main/kotlin/com/ophymx/chapbook/app/model/CatalogSession.kt
package com.ophymx.chapbook.app.model

import com.ophymx.chapbook.Catalog
import kotlinx.coroutines.asCoroutineDispatcher
import kotlinx.coroutines.withContext
import java.util.concurrent.Executors

/**
 * One catalog being browsed, on the one thread it is allowed to be on.
 *
 * Every [Catalog] call blocks and the handle is one thread's at a time,
 * so a session owns a thread and hops every call onto it. The catalog is
 * opened on first use — through the app, so it browses over the app's
 * transport and signs in through the app's credential store — and
 * closed on the thread at the end. [savedId] is 0 for a catalog that
 * has no saved row, a pasted URL.
 */
class CatalogSession(private val shelf: Shelf, private val savedId: Long) : AutoCloseable {
    private val executor = Executors.newSingleThreadExecutor { r -> Thread(r, "chapbook-catalog") }
    private val thread = executor.asCoroutineDispatcher()
    private var catalog: Catalog? = null

    suspend fun <T> use(block: Catalog.() -> T): T = withContext(thread) {
        val c = catalog
            ?: checkNotNull(shelf.withApp { browse(savedId) }) { "the catalog did not open; see logcat" }
                .also { catalog = it }
        c.block()
    }

    override fun close() {
        executor.execute {
            catalog?.close()
            catalog = null
        }
        executor.shutdown()
    }
}
